extensions/PhabBugz/lib/Util.pm

```diff
@@ -305,4 +306,135 @@ sub set_phab_user {
+  my $rev_identifier = 'D' . $revision->id;
+
+  INFO("$rev_identifier: Setting reviewer rotation.");
+
+  # Load a fresh version of the revision with Heralds changes.
+  $revision = Bugzilla::Extension::PhabBugz::Revision->new_from_query(
+    {phids => [$revision->phid]});
+
+  # 1. Find out what the project reviewers and individual reviewers are.
+  # If the revision has a blocking reviewer group set, normally is 1) blocking
+  # and 2) ends in "-reviewers". Normally Herald will set this if certain
+  # conditions are met. If a blocking reviewer group cannot be found then
+  # do nothing.
+  my $blocking_project;
+  my @blocking_users;
+
+  foreach my $reviewer (@{$revision->reviews}) {
+    next if !$reviewer->{is_blocking};    # Only interested in blocking
+    if ($reviewer->{is_project}) {
+      next if $reviewer->{user}->name !~ /-reviewers$/;    # Only interested in reviewer groups
+      $blocking_project = $reviewer->{user};
+    }
+    else {
+      push @blocking_users, $reviewer->{user};
+    }
+  }
+
+  if (!$blocking_project) {
+    INFO("$rev_identifier: Blocking reviewer project not found.");
+    return;
+  }
+
+  INFO( "$rev_identifier: Blocking reviewer project "
+      . $blocking_project->name
+      . ' found.');
+
+  # 2. Once the blocking reviewer group is determined, query Phabricator for
+  # list of group members and match up the BMO user account. Sort them by user
+  # id descending.
+  my $project_members
+    = [sort { $a->id <=> $b->id } @{$blocking_project->members}];
+
+  # 3. Make sure that none of the individual group members are not already
+  # set as a blocking reviewer. If so, then remove the blocking group and return.
+  foreach my $member (@{$project_members}) {
+    if (any { $_->id == $member->id } @blocking_users) {
+      INFO(
+        "$rev_identifier: Member of blocking reviewer project already set as a reviewer. Removing blocking reviewer project."
+      );
+      $revision->remove_reviewer($blocking_project->phid);
+      $revision->update;
+      return;
+    }
+  }
+
+  # 4. Going in order, look up in the phab_reviewer_rotation table for each
+  # user to see if they are already a reviewer on another attachment.
+  my $dbh = Bugzilla->dbh;
+
+  my $found_reviewer;
+  foreach my $member (@{$project_members}) {
+
+   # 5. If the user has a revison they are reviewing currently, load the revision
+   # details and check if it is closed. If it is, then clear the row from the table.
+    my $rev_phid = $dbh->selectrow_array(
+      'SELECT revision_phid FROM phab_reviewer_rotation WHERE project_phid = ? AND user_phid = ?',
+      undef, $blocking_project->phid, $member->phid
+    );
+
+    if ($rev_phid) {
+      INFO("$rev_identifier: Previous review found for " . $member->id);
+
+      my $rev_obj = Bugzilla::Extension::PhabBugz::Revision->new_from_query(
+        {phids => [$rev_phid]});
+
+# 6. If the user is already a reviewer or they were but their revision is now closed, skip to the next
```

Comment:
If their revision is now closed, couldn't they be the `$found_reviewer` for this revision? They no longer have a revision to review.

Reply:
The idea was that even though their revision is now closed and they are no longer reviewing it, I would still pick the next person in the list since the previous just reviewed something. This is to not favor any single person and make sure everyone gets a fair amount.